python/neutron-understack/neutron_understack/neutron_understack_mech.py

```diff
@@ -307,17 +327,42 @@ def bind_port(self, context: PortContext) -> None:
                     },
                 )
 
-    def check_segment(self, segment):
-        """Verify a segment is valid for the Understack MechanismDriver.
+    def _find_or_create_network_segment(self, context: PortContext):
+        """Ensure that a VLAN-type network segment is present for this network.
 
-        Verify the requested segment is supported by Understack and return True or
-        False to indicate this to callers.
+        In the VXLAN fabric, a Network has a VLAN created on every switch where
+        that network is in use.  The VLAN is local to the VLAN GROUP (pair of
+        switches) and must be assigned an arbitrary vlan number that is unique
+        within that VLAN GROUP.
+
+        We use the switch name to figure out which VLAN GROUP this network is
+        being bound to.  We then look for existing network segments in that VLAN
+        GROUP.  If none are found then we create a new one with a dynamically
+        allocated VLAN ID.
         """
-        network_type = segment[api.NETWORK_TYPE]
-        return network_type in [
-            p_const.TYPE_VXLAN,
-            p_const.TYPE_VLAN,
-        ]
+        switch_name = str(
+            context.current["binding:profile"]["local_link_information"][0][
+                "switch_info"
+            ]
+        )
+
+        vlan_group_name = vlan_group_name_convention.for_switch(switch_name)
+
+        for segment in context.network.network_segments:
+            if (
+                segment.get(api.NETWORK_TYPE) == p_const.TYPE_VLAN
+                and segment.get(api.PHYSICAL_NETWORK) == vlan_group_name
+            ):
+                LOG.info("Using existing VLAN network segment %s", segment)
+                return segment
+
+        segment_args = {
+            api.NETWORK_TYPE: p_const.TYPE_VLAN,
+            api.PHYSICAL_NETWORK: vlan_group_name,
+        }
+        segment = context.allocate_dynamic_segment(segment_args)
```

Comment:
We shouldn't have to do this part. The port will come in with a physical network request that should match a network-segment-range and we'll just bind on that.